### PR TITLE
neo: mint GAS after NEO transfer processing

### PR DESCRIPTION
**Pull request by roman-khimov**
See neo-project/neo#2734.

(it won't merge as of now, but it's needed for 0.98.5)

### CODE REVIEW

File: pkg/core/native/native_neo.go [hunk 1]
```diff
@@ -461,23 +467,17 @@ func (n *NEO) balanceFromBytes(si *state.StorageItem) (*big.Int, error) {
...
 	if err != nil {
-		return err
+		return nil, err
 	}
 	acc.BalanceHeight = ic.Block.Index
 
-	// Must store acc before GAS distribution to fix acc's BalanceHeight value in the storage for
-	// further acc's queries from `onNEP17Payment` if so, see https://github.com/nspcc-dev/neo-go/pull/2181.
-	key := makeAccountKey(h)
-	ic.DAO.PutStorageItem(n.ID, key, acc.Bytes())
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:477]: Why did you remove these lines? It's always `GetAndChange` in C# and two lines above we've changed acc's BalanceHeight. We must save account anyway here.

Comment (roman-khimov) [reply pkg/core/native/native_neo.go:477]: It's just no longer needed, it'll be saved before we're to invoke `Mint` in `(c *nep17TokenNative) updateAccBalance()` or `(c *nep17TokenNative) addTokens()`.

File: pkg/core/native/native_nep17.go [hunk 2]
```diff
@@ -128,7 +128,14 @@ func addrToStackItem(u *util.Uint160) stackitem.Item {
 }
 
 func (c *nep17TokenNative) postTransfer(ic *interop.Context, from, to *util.Uint160, amount *big.Int,
-	data stackitem.Item, callOnPayment bool) {
+	data stackitem.Item, callOnPayment bool, postCalls ...func()) {
+	defer func() {
```
Comment (AnnaShaleva) [inline pkg/core/native/native_nep17.go:132]: I'm not sure, but if an error is returned from OnNEP17Payment call then `defer` will still be invoked. Is it the desired behaviour? Although the transaction will be FAULTed and it's not very important.

File: pkg/core/native/native_neo.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go:477]: I still have doubds about this removal. It was here not only to save BalanceHeight, but also for the following case: if the subsequent code will return an error (e.g. `ModifyAccountVotes` inside `VoteInternal` will return an error), then no account will be saved in neo-go implementation (against the C# node where account's height is updated). NEO's `vote` doesn't panic if an error is returned from `VoteInternal`, instead it puts `false` on stack. Here comes the state diff.

Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go:477]: Although there's a *very* tiny chance that `ModifyAccountVotes` will return an error.

Comment (roman-khimov) [reply pkg/core/native/native_neo.go:477]: In this case the transfer will fail for sure and GAS shouldn't be minted, thus account height shouldn't be updated and DB can be left as is. Although for `VoteInternal` it can be solved.

Comment (roman-khimov) [reply pkg/core/native/native_neo.go:477]: Wow. A simple
```
--- a/pkg/core/native/native_neo.go
+++ b/pkg/core/native/native_neo.go
@@ -812,6 +812,7 @@ func (n *NEO) VoteInternal(ic *interop.Context, h util.Uint160, pub *keys.Public
        if err != nil {
                return err
        }
+       ic.DAO.PutStorageItem(n.ID, key, acc.Bytes()) // It won't be changed after that, but the DB must be updated.
        if err := n.ModifyAccountVotes(acc, ic.DAO, new(big.Int).Neg(&acc.Balance), false); err != nil {
                return err
        }
@@ -819,7 +820,6 @@ func (n *NEO) VoteInternal(ic *interop.Context, h util.Uint160, pub *keys.Public
        if err := n.ModifyAccountVotes(acc, ic.DAO, &acc.Balance, true); err != nil {
                return err
        }
-       ic.DAO.PutStorageItem(n.ID, key, acc.Bytes())
        if newGas != nil { // Can be if it was already distributed in the same block.
                n.GAS.mint(ic, h, newGas, true)
        }
```
Breaks some tests.

Comment (roman-khimov) [reply pkg/core/native/native_neo.go:477]: OK, NM, that's intended.

File: pkg/core/native/native_neo.go [hunk 3]
```diff
@@ -461,23 +467,17 @@ func (n *NEO) balanceFromBytes(si *state.StorageItem) (*big.Int, error) {
 	return &acc.Balance, err
 }
 
-func (n *NEO) distributeGas(ic *interop.Context, h util.Uint160, acc *state.NEOBalance) error {
+func (n *NEO) distributeGas(ic *interop.Context, h util.Uint160, acc *state.NEOBalance) (*big.Int, error) {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/native/native_neo.go]: Yep, removed.

File: pkg/core/native/native_nep17.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/core/native/native_nep17.go:133]: Fixed.

File: pkg/core/native/native_neo.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/native/native_neo.go:477]: > NEO's vote doesn't panic if an error is returned from VoteInternal,

I think where we have an error returned C# node will just `throw` and that will kill the execution.

Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go:477]: You're right, then we can safely remove it.

**Review by fyrchik — APPROVED** [review]

**Comment by fyrchik** [comment]
Just to clarify, what is the _expected_ transfer event sequence in this case (would be nice to have a compatibility test like this):
There are accounts A, B and contract C all with non-zero balance.
1. A transfers NEO to C
2. C transfers NEO and GAS (explicitly) to B in `OnNEP11Payment`.
